flytekit/configuration/plugin.py

```diff
@@ -0,0 +1,74 @@
+"""Defines a plugin API allowing other libraries to modify the behavior of flytekit.
+
+Libraries can register by defining an object that follows the same API as FlytekitPlugin
+and providing entry pont with the group name "flytekit.plugin". In `setuptools`,
+you can specific them with:
+
+```python
+setup(entry_points={
+    "flytekit.plugin": ["my_plugin=..."]
+})
+```
+
+or in pyproject.toml:
+
+```toml
+[project.entry-points."flytekit.plugin"]
+my_plugin = "..."
+```
+"""
+from typing import Optional
+
+from click import Command
+from importlib_metadata import entry_points
+
+from flytekit.configuration import Config, get_config_file
+from flytekit.loggers import cli_logger
+from flytekit.remote import FlyteRemote
+
+
+class FlytekitPlugin:
```

Comment:
I wonder if this is a good use case for [Protocols](https://docs.python.org/3/library/typing.html#typing.Protocol) so we can type-check the API of libraries that use this plugin. Maybe overkill for now

Reply:
If the API surface gets bigger, a protocol makes it easier to figure out what to implement. (Although, I hope it does not get too big)

I added the protocol in https://github.com/flyteorg/flytekit/pull/2039/commits/dde46d9126aa936b9ec74ec63fa9f140c6bdba79